src/actions/nodes.js

```diff
@@ -38,3 +41,52 @@ export function toggleNodeHovered(nodeHovered) {
     nodeHovered
   };
 }
+
+export const TOGGLE_NODE_DATA_LOADING = 'TOGGLE_NODE_DATA_LOADING';
+
+/**
+ * Toggle whether to display the loading spinner
+ * @param {boolean} loading True if pipeline is still loading
+ */
+export function toggleNodeDataLoading(loading) {
+  return {
+    type: TOGGLE_NODE_DATA_LOADING,
+    loading
+  };
+}
+
+export const ADD_NODE_METADATA = 'ADD_NODE_METADATA';
+
+/**
+ * Toggle whether to display the loading spinner
+ * @param {boolean} loading True if pipeline is still loading
+ */
+export function addNodeMetadata(data) {
+  return {
+    type: ADD_NODE_METADATA,
+    data
+  };
+}
+
+/**
+ * update node metadata on selection, loading new data if it has not been previously called
+ * @param {string} nodeID node id of clicked node
+ * @return {function} A promise that resolves when the data is loaded
+ */
+export function loadNodeData(nodeID) {
+  return async function(dispatch, getState) {
+    const { asyncDataSource, node } = getState();
+
+    dispatch(toggleNodeClicked(nodeID));
+
+    if (asyncDataSource) {
+      if (!node.fetched[nodeID]) {
+        dispatch(toggleNodeDataLoading(true));
+        const url = getUrl('nodes', nodeID);
+        const nodeData = await loadJsonData(url);
+        dispatch(addNodeMetadata({ id: nodeID, data: nodeData }));
```

Comment:
I don't think the nesting here is necessary. Also I would be inclined to have it pass `nodeID` instead of `id` - it's mostly a convention used across a lot of the app, and makes the reducer code a bit more clear. Hence you could do something like this:
```es6
dispatch(addNodeMetadata({ ...nodeData, nodeID }));
```